File: apps/pipeline/src/postthedoc/tokens.py
"""Signed tokens for email links. Same format as worker/src/tokens.ts.

token = base64url("{purpose}.{user_id}.{version}.{exp}") + "." + base64url(HMAC-SHA256)
exp = 0 means the token never expires.
"""

import base64
import hashlib
import hmac
import time
from dataclasses import dataclass


@dataclass(frozen=True)
class TokenData:
    purpose: str
    user_id: str
    version: int
    exp: int


def _b64(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _unb64(data: str) -> bytes:
    return base64.urlsafe_b64decode(data + "=" * (-len(data) % 4))


def _mac(secret: str, payload: str) -> bytes:
    return hmac.new(secret.encode(), payload.encode(), hashlib.sha256).digest()


def sign(secret: str, purpose: str, user_id: str, version: int, ttl: int | None = None) -> str:
    exp = int(time.time()) + ttl if ttl else 0
    payload = f"{purpose}.{user_id}.{version}.{exp}"
    return f"{_b64(payload.encode())}.{_b64(_mac(secret, payload))}"


def verify(secret: str, token: str, purposes: set[str]) -> TokenData | None:
    try:
        encoded, signature = token.split(".")
        payload = _unb64(encoded).decode()
        if not hmac.compare_digest(_unb64(signature), _mac(secret, payload)):
            return None
        purpose, user_id, version, exp = payload.split(".")
        data = TokenData(purpose, user_id, int(version), int(exp))
    except ValueError:
        return None
    if data.purpose not in purposes or (data.exp and data.exp < time.time()):
        return None
    return data

File: apps/worker/src/tokens.ts
// Signed tokens for email links. Same format as src/postthedoc/tokens.py:
// base64url("{purpose}.{userId}.{version}.{exp}") + "." + base64url(HMAC-SHA256), exp = 0 never expires.

export type Purpose = "confirm" | "manage" | "unsubscribe";

export interface TokenData {
  purpose: Purpose;
  userId: string;
  version: number;
  exp: number;
}

const encoder = new TextEncoder();

function toB64url(bytes: Uint8Array): string {
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function fromB64url(value: string): Uint8Array {
  const b64 = value.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(b64 + "=".repeat((4 - (b64.length % 4)) % 4));
  return Uint8Array.from(binary, (c) => c.charCodeAt(0));
}

function hmacKey(secret: string): Promise<CryptoKey> {
  return crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign", "verify"],
  );
}

export async function sign(
  secret: string,
  purpose: Purpose,
  userId: string,
  version: number,
  ttlSeconds?: number,
): Promise<string> {
  const exp = ttlSeconds ? Math.floor(Date.now() / 1000) + ttlSeconds : 0;
  const payload = `${purpose}.${userId}.${version}.${exp}`;
  const mac = await crypto.subtle.sign("HMAC", await hmacKey(secret), encoder.encode(payload));
  return `${toB64url(encoder.encode(payload))}.${toB64url(new Uint8Array(mac))}`;
}

export async function verify(
  secret: string,
  token: string,
  purposes: Purpose[],
): Promise<TokenData | null> {
  try {
    const [encoded, signature, ...rest] = token.split(".");
    if (!encoded || !signature || rest.length) return null;
    const payloadBytes = fromB64url(encoded);
    const valid = await crypto.subtle.verify(
      "HMAC",
      await hmacKey(secret),
      fromB64url(signature),
      payloadBytes,
    );
    if (!valid) return null;

    const [purpose, userId, version, exp, ...extra] = new TextDecoder()
      .decode(payloadBytes)
      .split(".");
    if (userId === undefined || version === undefined || exp === undefined || extra.length > 0) {
      return null;
    }
    const data: TokenData = {
      purpose: purpose as Purpose,
      userId,
      version: Number(version),
      exp: Number(exp),
    };
    if (!purposes.includes(data.purpose) || !Number.isInteger(data.version)) return null;
    if (data.exp && data.exp < Date.now() / 1000) return null;
    return data;
  } catch {
    return null; // invalid base64
  }
}
